# Vector Operations

I will be using the `cereal.csv` file, which was used in Chapter 3, as my data source. 


In [1]:
# import libraries
import pandas as pd
import numpy as np # this is used for numerical computing
import numpy.typing as npt # this is used for type annotation

In [2]:
# load data into dataframe
df = pd.read_csv('../datasets/cereal.csv')


Before I begin, I'd like to explain what vectors are.

Vectors are ordered lists of numbers used to represent quantities or data mathematically. 

Vectors are usually represented using NumPy arrays. (_Note that this does not mean all arrays are vectors_)

In this project, I will represent selected columns as vectors and perform vector operations on them.

In [3]:
# represent the 'sugars' and 'calories' columns as  vectors
sugar_vector = np.array(df['sugars'])
calories_vector = np.array(df['calories'])

# check that they are both arrays
print(f"Sugar Vector is a {type(sugar_vector)}")
print(f"Calories Vector is a {type(calories_vector)}")

Sugar Vector is a <class 'numpy.ndarray'>
Calories Vector is a <class 'numpy.ndarray'>


The first question I'll be asking is: "Do cereals that have more sugar also tend to have more calories?"

To answer this question, I will be using Cosine Similarity

In [4]:
# create function to calculate cosine similarity
def cosine_similarity(A: npt.NDArray[np.float64], B: npt.NDArray[np.float64]) -> float:
    # compute dot product
    dot_prod = np.dot(A, B)

    # compute the norm of the vectors
    norm_A = np.linalg.norm(A)
    norm_B = np.linalg.norm(B)

    # find the product of the norms
    vector_norm = norm_A * norm_B

    # compute cosine similarity
    cos_sim = np.round(dot_prod / vector_norm, 3)

    return(cos_sim)



In [5]:
# print cosine similarity
cos_sim = cosine_similarity(sugar_vector, calories_vector)

print(f"The Cosine Similarity of the Sugar column and Calories column is {cos_sim}")

The Cosine Similarity of the Sugar column and Calories column is 0.883


_What does Cosine Similarity even mean?_

Cosine Similarity measures how similar two data items are by calculating the cosine of the angle between them. It focuses on the direction rather than magnitude, with a score of 1 meaning identical, 0 unrelated, and -1 opposite.


The Cosine Similarity of `0.883` means that there is a strong positive similarity between the two vectors.

This may suggest that cereals with higher sugar content generally have higher calorie values.


**NOTE:** 

Cosine Similarity does not measure correlation. Cosine similarity measures geometric similarity, while correlation measures statistical relationship.



**We have answered the question!**

The cosine similarity of 0.883 suggests that the sugar and calorie vectors point in a similar direction, meaning cereals with higher sugar values often also have higher calorie values in this dataset.

**Onto the next!!**

Does Raisin Bran contain more calories and sugar than Raisin Bran Nut?

Let's compute the difference in nutrients.

In [ ]:
def vector_difference(cereal_1: str, cereal_2: str, nutrients: list[str]) -> dict[str, float]:
    """Calculates the difference between two vectors. 
    
    A positive value means cereal_1 contains more of that nutrient.
    A negative value means cereal_2 contains more of that nutrient.

    """

    # store original cereal names
    cereal_1_name = cereal_1.title()
    cereal_2_name = cereal_2.title()
    
    # filter data for selected cereals
    cereal_1_df = df[df["name"] == cereal_1_name][nutrients]
    cereal_2_df = df[df["name"] == cereal_2_name][nutrients]

    # check if cereal inputed exists
    if cereal_1_df.empty:
        raise ValueError(f"{cereal_1_name} is not found in the dataset")
    
    if cereal_2_df.empty:
        raise ValueError(f"{cereal_2_name} is not found in the dataset")
    
    # convert dataframes to Numpy array
    cereal_1 = cereal_1_df.to_numpy().astype(float)
    cereal_2 = cereal_2_df.to_numpy().astype(float)

    # calculate difference in nutrients
    diff = cereal_1 - cereal_2

    # create dictionary that takes nutrient as key and difference as value
    summary = {n : float(d) for n, d in zip(nutrients, diff[0])}
    return summary

    

Now that we have our function, let's calculate the difference in calories and sugar between Raisin Bran and Raisin Nut Bran.

In [83]:
print(vector_difference(cereal_1='Raisin Bran', cereal_2='Raisin Nut Bran', nutrients=['calories', 'sugars']))

{'calories': 20.0, 'sugars': 4.0}


The result shows that Raisin Bran contains 20 more calories and 4 grams more sugar than Raisin Nut Bran in this dataset.